src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,665 @@
+    /// The Client ID for the connection
+    client_id: u64,
+}
+
+// TODO: Add markable id for this client
+impl<P> Debug for RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RpcClientConnectionInner")
+            .field("client id", &self.client_id)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<P> RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// Create a new connection.
+    pub fn new(stream: TcpStream, timeout_options: ClientTimeoutOptions, client_id: u64) -> Self {
+        let task_timeout = timeout_options.task_timeout.as_secs();
+        Self {
+            stream: UnsafeCell::new(stream),
+            timeout_options,
+            seq: AtomicU64::new(0),
+            clock_instant: Instant::now(),
+            received_keepalive_timestamp: AtomicU64::new(0),
+            packets_keeper: PacketsKeeper::new(task_timeout),
+            resp_buf: UnsafeCell::new(BytesMut::with_capacity(
+                common::DEFAULT_TCP_RESPONSE_BUFFER_SIZE,
+            )),
+            req_buf: UnsafeCell::new(BytesMut::with_capacity(
+                common::DEFAULT_TCP_REQUEST_BUFFER_SIZE,
+            )),
+            client_id,
+        }
+    }
+
+    /// Recv request header from the stream
+    pub async fn recv_header(&self) -> Result<RespHeader, RpcError> {
+        // Try to read to buffer
+        match self.recv_len(REQ_HEADER_SIZE).await {
+            Ok(()) => {}
+            Err(err) => {
+                debug!("Failed to receive request header: {:?}", err);
+                return Err(err);
+            }
+        }
+
+        let buffer: &mut BytesMut = unsafe { &mut *self.resp_buf.get() };
+        let req_header = RespHeader::decode(buffer)?;
+        debug!("Received response header: {:?}", req_header);
+
+        Ok(req_header)
+    }
+
+    /// Receive response body from the server.
+    pub async fn recv_len(&self, len: u64) -> Result<(), RpcError> {
+        let mut req_buffer: &mut BytesMut = unsafe { &mut *self.resp_buf.get() };
+        req_buffer.resize(u64_to_usize(len), 0);
+        let reader = self.get_stream_mut();
+        match read_exact_timeout!(reader, &mut req_buffer, self.timeout_options.read_timeout).await
+        {
+            Ok(size) => {
+                debug!("{:?} Received response body len: {:?}", self, size);
+                Ok(())
+            }
+            Err(err) => {
+                debug!("{:?} Failed to receive response header: {:?}", self, err);
+                Err(RpcError::InternalError(err.to_string()))
+            }
+        }
+    }
+
+    /// Send a request to the server.
+    /// We need to make sure
+    /// Current send packet need to be in one task, so if we need to send ping and packet
+    /// we need to make sure only one operation is running, like select.
+    pub async fn send_data(
```

Comment:
Please remove all the `pub` in methods of `RpcClientConnectionInner`

Reply:
OK.